### PR TITLE
MSVC port [ESD-1116]

### PR DESCRIPTION
**Pull request by pmiettinen**
## Main changes
- Use alloca instead of Variable-Length-Arrays
- Subchange to previous: Compiler switch to use heap instead of stack (can be removed if deemed never gonna need)
- Add travis jobs: GCC-4.8, MSVC (32/64 bit environments), OSX AppleClang
- Remove static array indices
- In log macros require one mandatory argument, otherwise MSVC variadic macro trailing comma removal works incorrectly.
- Size of `void *` is not known; cast to `char *` or use `uintptr_t`. GCC has implicitly working extension for this.
- `memcpy_s` is available on some platform and others not. Try to navigate trough these. I think this should be worked further in a follow-up PR. Most likely remove our implementation and see if `STDC_LIB_EXT1` support is available on all our GCC based platforms. `STDC_LIB_EXT1` support provides `memcpy_s`.
- Add dynamic linking export / import semantics

## Testing
- HITL https://github.com/swift-nav/piksi_firmware_private/pull/2752

### CODE REVIEW

File: src/single_epoch_solver.c [hunk 1]
```diff
@@ -482,11 +487,11 @@ static s8 pvt_solve(const u8 n_used,
     /* Construct a geometry matrix.  Each row (satellite) is
      * independently normalized into a unit vector. */
     for (u8 i = 0; i < 3; i++) {
-      G[j][i] = -los[i];
+      *(G + j + i) = -los[i];
```
Comment (pmiettinen) [inline src/single_epoch_solver.c]: incorrect..

**Comment by nsirola** [comment]
Looks good to me. Also compiles cleanly in `gnss-converters` and passes `valgrind` there

File: include/swiftnav/common.h [hunk 2]
```diff
@@ -13,11 +13,53 @@
 #ifndef LIBSWIFTNAV_COMMON_H
 #define LIBSWIFTNAV_COMMON_H
 
+#include <assert.h>
+#include <stdint.h>
+
+#ifdef LSN_USE_HEAP
+#include <stdlib.h>
+#define LSN_ALLOCATE(size) malloc(size)
```
Comment (benjaminaltieri) [inline include/swiftnav/common.h]: I think eventually I'd prefer to see a generalized module that allows a user to define their own allocation (i.e. using RTOS mempool or other custom management) but defaulting to stack allocation like this with the option of heap allocation.

File: src/single_epoch_solver.c [hunk 3]
```diff
@@ -509,6 +516,8 @@ static s8 pvt_solve(const u8 n_used,
                               (double *)lsq_data->V);
   if (ret < 0) {
     log_warn("Under-determined system, n_used = %u", n_used);
+    LSN_FREE(G);
```
Comment (benjaminaltieri) [inline src/single_epoch_solver.c]: not a huge fan of peppering in duplicate stacks of `LSN_FREE` calls throughout to support multiple return in the functions. I know `goto error` is a controversial choice, so if not that then single return?

Comment (nsirola) [reply src/single_epoch_solver.c]: This function is only ever called from `pvt_iter()`, so one option would be to allocate and free the working arrays there and pass pointers.

Comment (pmiettinen) [reply src/single_epoch_solver.c]: I made a refactor so that there's one deallocation point per function.

**Comment by pmiettinen** [comment]
@benjaminaltieri - No need for changes has been detected in the consumer repositories.

**Comment by nsirola** [comment]
Lgtm.

**Comment by benjaminaltieri** [comment]
@anth-cole any reason we should hold on merging this or is it good to go?

**Comment by anth-cole** [comment]
no need to wait AFAIK

**Comment by pmiettinen** [comment]
LSN-Starling-PFWP all CI checks showing green -> merging.

**Pull request merged**